bgpd/bgp_mplsvpn.c

```diff
@@ -1257,29 +1515,135 @@ void vpn_leak_from_vrf_update(struct bgp *to_bgp,	     /* to */
+				arg_bits_length = chunk->argument_bits_length;
+			transposition_offset =
+				block_bits_length + node_bits_length;
+			if (chunk->function_bits_length)
+				transposition_length =
+					chunk->function_bits_length;
+		}
+
 		encode_label(
 			from_bgp->vpn_policy[afi].tovpn_sid_transpose_label,
 			&label);
 		static_attr.srv6_l3vpn = XCALLOC(MTYPE_BGP_SRV6_L3VPN,
 				sizeof(struct bgp_attr_srv6_l3vpn));
 		static_attr.srv6_l3vpn->sid_flags = 0x00;
 		static_attr.srv6_l3vpn->endpoint_behavior = 0xffff;
-		static_attr.srv6_l3vpn->loc_block_len =
-			BGP_PREFIX_SID_SRV6_LOCATOR_BLOCK_LENGTH;
-		static_attr.srv6_l3vpn->loc_node_len =
-			BGP_PREFIX_SID_SRV6_LOCATOR_NODE_LENGTH;
-		static_attr.srv6_l3vpn->func_len =
-			BGP_PREFIX_SID_SRV6_FUNCTION_LENGTH;
-		static_attr.srv6_l3vpn->arg_len =
-			BGP_PREFIX_SID_SRV6_ARGUMENT_LENGTH;
+		static_attr.srv6_l3vpn->loc_block_len = block_bits_length;
+		static_attr.srv6_l3vpn->loc_node_len = node_bits_length;
+		static_attr.srv6_l3vpn->func_len = func_bits_length;
+		static_attr.srv6_l3vpn->arg_len = arg_bits_length;
 		static_attr.srv6_l3vpn->transposition_len =
-			BGP_PREFIX_SID_SRV6_TRANSPOSITION_LENGTH;
+			transposition_length;
 		static_attr.srv6_l3vpn->transposition_offset =
-			BGP_PREFIX_SID_SRV6_TRANSPOSITION_OFFSET;
+			transposition_offset;
 		memcpy(&static_attr.srv6_l3vpn->sid,
 		       from_bgp->vpn_policy[afi].tovpn_sid_locator,
 		       sizeof(struct in6_addr));
+	} else if (from_bgp->tovpn_sid_locator) { /* per-VRF SID */
+		struct listnode *node;
+		struct srv6_locator_chunk *chunk;
+		uint8_t block_bits_length =
+			BGP_PREFIX_SID_SRV6_LOCATOR_BLOCK_LENGTH;
+		uint8_t node_bits_length =
+			BGP_PREFIX_SID_SRV6_LOCATOR_NODE_LENGTH;
+		uint8_t func_bits_length = BGP_PREFIX_SID_SRV6_FUNCTION_LENGTH;
+		uint8_t arg_bits_length = BGP_PREFIX_SID_SRV6_ARGUMENT_LENGTH;
+		uint8_t transposition_offset =
+			BGP_PREFIX_SID_SRV6_TRANSPOSITION_OFFSET;
+		uint8_t transposition_length =
+			BGP_PREFIX_SID_SRV6_TRANSPOSITION_LENGTH;
+		bool found = false;
+
+		/* Perform SRv6 locator lookup */
+		if (to_bgp->srv6_locator_chunks) {
+			for (ALL_LIST_ELEMENTS_RO(to_bgp->srv6_locator_chunks,
+						  node, chunk)) {
+				if (sid_same(from_bgp->tovpn_sid_locator,
+					     &chunk->prefix.prefix)) {
+					found = true;
+					break;
+				}
+			}
+		}
+
+		/*
+		 * If the SRv6 locator exists, we use the bits length parameters
+		 * of the SRv6 locator instead of using the default values
+		 */
+		if (found) {
+			if (chunk->block_bits_length)
+				block_bits_length = chunk->block_bits_length;
+			if (chunk->node_bits_length)
+				node_bits_length = chunk->node_bits_length;
+			if (chunk->function_bits_length)
+				func_bits_length = chunk->function_bits_length;
+			if (chunk->argument_bits_length)
+				arg_bits_length = chunk->argument_bits_length;
+			transposition_offset =
+				block_bits_length + node_bits_length;
+			if (chunk->function_bits_length)
+				transposition_length =
```

Comment:
If the user/operator configures function length = 32 bits, then "transposition_length" field here will be set to 32, which is wrong. This is because, the number of bits that can be transposed from the function part of SID into label field (label field is 24 bits) of NLRI is 20bits only. (20 bits in case of VPNv4/v6 and the remaining 4 bits in label is used for BOS bit and EXP bits). So if the function_bits_length is > 20 bits, then set the transposition_length to 20 bits and add the rest of the bits i.e, function_bits_length - 20 to transposition offset.

From section 3.2.1 of RFC 9252:
  "As an example, consider that the sum of the Locator Block and the
   Locator Node parts is 64.  For an SRv6 SID where the entire Function
   part of size 16 bits is transposed, the transposition offset is set
   to 64 and the transposition length is set to 16.  **While for an SRv6
   SID for which the FL is 24 bits and only the lower order 20 bits are
   transposed (e.g., due to the limit of the MPLS Label field size), the
   transposition offset is set to 68 and the transposition length is set
   to 20."**

Reply:
In addition, BGP needs to validate the that the transpose length is not greater than the label field when processing  type 5 prefix-SID attribute (BGP_PREFIX_SID_SRV6_L3_SERVICE). In case the transpose_length is > 20 bits (in case of VPN) then BGP needs to mark this path as ineligible for bestpath calculation. May be this can be done  separately in the future, but just putting it out there. 

According to "error handling" section of RFC 9252:
"The SRv6 SID value in the SRv6 SID Information Sub-TLV is invalid
   when the SID Structure Sub-Sub-TLV transposition length is greater
   than the number of bits of the label field or if any of the
   conditions for the fields of the Sub-Sub-TLV, as specified in
   [Section 3.2.1](https://datatracker.ietf.org/doc/html/rfc9252#section-3.2.1), is not met.  The transposition offset and length MUST
   be 0 when the Sub-Sub-TLV is advertised along with routes where the
   Transposition Scheme is not applicable (e.g., for global IPv6 service
   [[RFC2545](https://datatracker.ietf.org/doc/html/rfc2545)] where there is no label field).  The path having any such
   Prefix-SID attribute without any valid SRv6 SID information MUST be
   considered ineligible during the selection of the best path for the
   corresponding prefix."

Reply:
> If the user/operator configures function length = 32 bits, then "transposition_length" field here will be set to 32, which is wrong. This is because, the number of bits that can be transposed from the function part of SID into label field (label field is 24 bits) of NLRI is upto 20bits only. (upto 20 bits in case of VPNv4/v6 since the remaining 4 bits in label is used for BOS bit and EXP bits). So if the function_bits_length is > 20 bits, then set the transposition_length to 20 bits and add the rest of the bits i.e, function_bits_length - 20 to transposition offset.
> 
> From section 3.2.1 of RFC 9252: "As an example, consider that the sum of the Locator Block and the Locator Node parts is 64. For an SRv6 SID where the entire Function part of size 16 bits is transposed, the transposition offset is set to 64 and the transposition length is set to 16. **While for an SRv6 SID for which the FL is 24 bits and only the lower order 20 bits are transposed (e.g., due to the limit of the MPLS Label field size), the transposition offset is set to 68 and the transposition length is set to 20."**

@Pdoijode Thanks for the suggestion. In my opinion, we should implement this feature/fix in a separate PR.

Note that in the current implementation of bgpd, this should not be a problem because the Function part of an SRv6 SID is limited to 8 bits (and it never exceeds 20 bits).

In bgpd, you can allocate SIDs in two ways:
* manually by providing a value in the range 1-255 that will be used as SRv6 SID Function (i.e., CLI command `sid vpn export (1-255)`)
* automatically, leaving the choice of the SID Function to bgpd (i.e., CLI command `sid vpn export auto`).

In the former case, the value that you provide cannot exceed 8 bits.
In the second case, bgpd chooses the value to use for the SRv6 SID Function in the range 1-255, and again this value never exceeds 8 bits.

So, in the current implementation, the SRv6 SID Function is limited to 8 bits, which is too limiting. We definitely need to remove this limitation.

I am already working to support SID functions larger than 8 bits, but this requires several changes to bgpd and to the bgpd CLI. I think adding these changes to this PR would be confusing and would make the review more difficult.

For this reason, I prefer to implement these changes in a separate PR. I will submit the new PR after this PR is merged. In the new PR I will also take into account that the MPLS label field is 20 bits long and will add code to handle the case you highlighted.

> In addition, BGP needs to validate the that the transpose length is not greater than the label field when processing type 5 prefix-SID attribute (BGP_PREFIX_SID_SRV6_L3_SERVICE). In case the transpose_length is > 20 bits (in case of VPN) then BGP needs to mark this path as ineligible for bestpath calculation. May be this can be done separately in the future, but just putting it out there.
> 
> According to "error handling" section of RFC 9252: "The SRv6 SID value in the SRv6 SID Information Sub-TLV is invalid when the SID Structure Sub-Sub-TLV transposition length is greater than the number of bits of the label field or if any of the conditions for the fields of the Sub-Sub-TLV, as specified in [Section 3.2.1](https://datatracker.ietf.org/doc/html/rfc9252#section-3.2.1), is not met. The transposition offset and length MUST be 0 when the Sub-Sub-TLV is advertised along with routes where the Transposition Scheme is not applicable (e.g., for global IPv6 service [[RFC2545](https://datatracker.ietf.org/doc/html/rfc2545)] where there is no label field). The path having any such Prefix-SID attribute without any valid SRv6 SID information MUST be considered ineligible during the selection of the best path for the corresponding prefix."

Yeah, that's an interesting feature. Again, in my opinion, this should be done in a separate PR.

Reply:
@riw777 We can add a check to this PR to return fail when the function length is greater than 20 bits. Then, I will open a new PR to remove this limitation and add support for the larger function length.